web/src/components/Header.js

```diff
@@ -6,37 +8,103 @@ import { connect } from 'react-redux';
 import { Link } from 'react-router-dom';
 import { push } from 'connected-react-router';
 
-
+import { getIsAdmin } from '../reducers/user';
 import { t } from '../i18n';
 
+import Icon, { faChevronDown, faChevronLeft, faEdit, faSignOutAlt } from './Icons';
+
 class Header extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      ariaExpanded: props.ariaExpanded,
+
+    };
+    this.handleOutsideClick = this.handleOutsideClick.bind(this);
+  }
+
+  componentDidMount() {
+    document.addEventListener('mousedown', this.handleOutsideClick);
+  }
+
+  handleOutsideClick = e => {
+    if (this.node.contains(e.target)) {
+      return;
+    }
+    this.setState({ ariaExpanded: false })
+  };
```

Comment:
This method and the `document.addEventListener` in `componentDidMount` seem like they ought to be tested, but I can't seem to figure out the right way to stub things and I don't want to hold up this PR any longer while I bang my head against the wall.

Reply:
![i got you](https://media.giphy.com/media/xULW8wR2MiuOdWUswg/giphy.gif)

### For clicking directly on the toggly bit

In the `opens and closes the dropdown when clicked` test, add this after you create the component:
```js
jest.spyOn(document, 'addEventListener');
jest.spyOn(document, 'removeEventListener');
```

Then after the first click simulation, add these assertions:
```js
expect(document.addEventListener).toHaveBeenCalled();
expect(document.addEventListener.mock.calls[0][0]).toEqual('click');
const handler = document.addEventListener.mock.calls[0][1];
```

And after the second click simulation, these assertions:
```js
expect(document.removeEventListener).toHaveBeenCalled();
expect(document.removeEventListener.mock.calls[0][0]).toEqual('click');
expect(document.removeEventListener.mock.calls[0][1]).toEqual(handler);
```

### For testing dom events

```js
it('handles outside click events', () => {
  // Create an instance of the component object but don't mount it. This
  // breaks us out of all the React requirements, which is handy because
  // we don't actually need them for these tests.
  const component = new Header({
    authenticated: true,
    currentUser: {
      role: 'admin',
      state: { id: 'wa', name: 'Washington' },
      username: 'frasiercrane@kacl.com'
    },
    isAdmin: true,
    pushRoute: () => {},
    ariaExpanded: false,
    showSiteTitle: true
  });

  // React sets the node for us, so since we're not Reacting, do it manually.
  // In the first case, make the contains fn return false.
  component.node = { current: { contains: jest.fn(() => false) } };

  jest.spyOn(document, 'removeEventListener');
  // By default, spyOn calls the real method as well, but that'll cause a
  // React error here because the component isn't mounted
  jest.spyOn(component, 'setState').mockImplementation(() => {});

  // Call the method under test
  component.handleOutsideClick({});

  // Make sure the state was set and the event handler was removed.
  expect(component.setState).toHaveBeenCalled();
  expect(component.setState.mock.calls[0][0]).toEqual({
    ariaExpanded: false
  });
  expect(document.removeEventListener).toHaveBeenCalled();
  expect(document.removeEventListener.mock.calls[0][0]).toEqual('click');
  // we can't check the function because of weird binding

  // Reset the mocks
  document.removeEventListener.mockClear();
  component.setState.mockClear();

  // Change the node contains function to return true this time.
  component.node.current.contains = jest.fn(() => true);

  component.handleOutsideClick({});

  expect(component.setState).not.toHaveBeenCalled();
  expect(document.removeEventListener).not.toHaveBeenCalled();
});
```